apis/r/R/SOMASparseNDArrayRead.R

```diff
@@ -24,7 +24,20 @@ SOMASparseNDArrayRead <- R6::R6Class(
     #' @return \link{SparseReadIter}
     sparse_matrix = function(zero_based=FALSE) {
       #TODO implement zero_based argument, currently doesn't do anything
-      SparseReadIter$new(sr = private$sr, shape = private$shape, zero_based=zero_based)
+
+
+      if (any(private$shape > .Machine$integer.max)) {
+        warning(
```

Comment:
Should we just error out here?  A warning and modifying the expected output is error-prone if the user is not paying close attention

Reply:
I struggled with this question but if we error out here we preclude the creation of sparse matrices from arrays with domains >= 2^31-1, even when all non-zero elements are an within R-friendly range of `[0, 2^31-1)`. To me this seemed like a reasonable trade-off: we can still create sparse matrices from these arrays, but we warn the user the resulting matrix's shape will be truncated to `c(2^31-1, 2^31-1)`. In the unlikely scenario that the array (& query) contains non-zero elements at indices >= 2^31-1 (or more than than 2^32-1 values), then we error out and suggest they access the data as an arrow table.

What do you think?

Reply:
Is there a performant way to find the max _used index_ and check if _that_ is > 2^31-1?

Reply:
Currently we check the coordinates of the arrow table before attempting to construct a sparse matrix. We could potentially use non-empty domain beforehand but we'll have an easier option in place after we starting the bounding box metadata.

Reply:
@aaronwolen this is awesome!!  Please either create a tracking task, or, make a note on #1445 so we remember

Reply:
> Thanks, @pablo-gar. I struggled with this question but if we error out here we preclude the creation of sparse matrices from arrays with domains >= 2^31-1, even when all non-zero elements are an within R-friendly range of [0, 2^31-1). 

Fair. 

>  In the unlikely scenario that the array (& query) contains non-zero elements at indices >= 2^31-1 (or more than than 2^32-1 values), then we error out and suggest they access the data as an arrow table.

And great solution

> > is there a performant way in R to find the max used index and check if that is > 2^31-1?
> We could potentially use non-empty domain. 

[I tried this already](https://github.com/single-cell-data/TileDB-SOMA/pull/1274#discussion_r1192742716) and decided not to go with it and to make the shape representation a bona-fide representation of the sparse shape and for parity with Python. I agree with @aaronwolen that once the bounding-box design is finalized we can revisit (and revisit in Python as well)